# 01 Importing libraries and dataset

In [1]:
#Importing libraries

import pandas as pd
import numpy as np
import os

In [2]:
#Creating path
path = r'C:\Users\zunai\OneDrive\Documents\Data Analytics\Achivement 4\11-22 Instacart Basket Analysis'

In [3]:
#Importing data sets
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared data', 'orders_products_merged_update2.pkl'))

In [4]:
df_cust = pd.read_csv(os.path.join(path, '02 Data', 'Original data', 'customers.csv'), index_col = False)

In [5]:
#Check
df_ords_prods_merge.shape

(32404859, 24)

In [6]:
df_ords_prods_merge.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                   999.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...    price_range_loc     Busiest day  Busiest days  \
0         Soda  ...  Mid-range product  Regularly busy  Regular days   
1         Soda  ...  Mid-range product  Regularly busy  Slowest days   
2         Soda  ...  Mid-range product  Regularly busy  Slowest days   
3         Soda  ...  Mid-range product      Least busy  Slowest days   
4         Soda  ...  Mid-range product      Least busy  Slowest days   

  Busiest_hours_of_day max_order  loyalty_flag average_price spender_flag  \
0          Most orders        10  New customer      6.367797  Low spender   
1       Average orders        10  New customer      6.367797  Low spender   
2          Most orders        10  New customer      6.367797  Low spender   
3       Average orders        10  New customer      6.367797  Low spender   
4          Most orders        10  New customer      6.367797  Low spender   

   median_days_since_prior_order         frequency_flag  
0                           21.0  Non-frequent customer  
1                           21.0  Non-frequent customer  
2                           21.0  Non-frequent customer  
3                           21.0  Non-frequent customer  
4                           21.0  Non-frequent customer  

[5 rows x 24 columns]

In [7]:
df_cust.shape

(206209, 10)

In [8]:
df_cust.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

# 02 Data wrangling

### Re-naming columns

In [9]:
df_cust.rename(columns= {'First Name' : 'first_name', 'Surnam' : 'surname','Gender' : 'gender','STATE' : 'state',
                          'Age' : 'age', 'n_dependants': 'no_of_dependants', 'fam_status' : 'marital_status'}, inplace = True)

In [10]:
df_cust.head()

user_id first_name   surname  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   no_of_dependants marital_status  income  
0                 3        married  165665  
1                 0         single   59285  
2                 2        married   99568  
3                 0         single   42049  
4                 1        married   40374

# 03 Consistency checks

In [12]:
#Descriptive analysis
df_cust.describe()

user_id            age  no_of_dependants         income
count  206209.000000  206209.000000     206209.000000  206209.000000
mean   103105.000000      49.501646          1.499823   94632.852548
std     59527.555167      18.480962          1.118433   42473.786988
min         1.000000      18.000000          0.000000   25903.000000
25%     51553.000000      33.000000          0.000000   59874.000000
50%    103105.000000      49.000000          1.000000   93547.000000
75%    154657.000000      66.000000          3.000000  124244.000000
max    206209.000000      81.000000          3.000000  593901.000000

In [13]:
#Check data types
df_cust.dtypes

user_id              int64
first_name          object
surname             object
gender              object
state               object
age                  int64
date_joined         object
no_of_dependants     int64
marital_status      object
income               int64
dtype: object

In [14]:
#Check for duplicates
df_cust[df_cust.duplicated()]

Empty DataFrame
Columns: [user_id, first_name, surname, gender, state, age, date_joined, no_of_dependants, marital_status, income]
Index: []

#### No duplicates found

In [16]:
#Check for missing values
df_cust.isnull().sum()

user_id                 0
first_name          11259
surname                 0
gender                  0
state                   0
age                     0
date_joined             0
no_of_dependants        0
marital_status          0
income                  0
dtype: int64

In [17]:
#View missing data
df_cust[df_cust['first_name'].isnull()==True]

user_id first_name   surname  gender         state  age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        no_of_dependants marital_status  income  
53                     2        married   41709  
73                     0         single   82518  
82                     3        married  117099  
99                     1        married  155673  
105                    3        married  151819  
...                  ...            ...     ...  
206038                 3        married   87783  
206044                 2        married  108488  
206090                 1        married   45275  
206162                 0         single   56325  
206171                 2        married   59222  

[11259 rows x 10 columns]

#### There are 11259 missing values for the first_name column (~5% of the data set). Although we could potentially remove the missing data without having too much of an impact on the outcome of our analysis, I will leave the missing values as we don't need the first_name to identify customers as we have the user_id column to identify them instead.

In [18]:
#Checking for columns with mixed data types
for col in df_cust.columns.tolist():
  weird = (df_cust[[col]].applymap(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_cust[weird]) > 0:
    print (col)

first_name


#### First_name column has a mixed data type

In [19]:
#Check the data type of 'first_name'
df_cust['first_name'].dtype

dtype('O')

In [20]:
#Convert first_name data type to string
df_cust['first_name'] = df_cust['first_name'].astype('str')

In [21]:
#Re-run check for columns with mixed data types
for col in df_cust.columns.tolist():
  weird = (df_cust[[col]].applymap(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_cust[weird]) > 0:
    print (col)

# 04 Reduce data types to save RAM 

In [22]:
df_cust['age'] = df_cust['age'].astype('int8')

In [23]:
df_cust['no_of_dependants'] = df_cust['no_of_dependants'].astype('int8')

In [24]:
df_cust['income'] = df_cust['income'].astype('int32')

In [25]:
df_cust['user_id'] = df_cust['user_id'].astype('str')

In [26]:
#Check for updated data types
df_cust.dtypes

user_id             object
first_name          object
surname             object
gender              object
state               object
age                   int8
date_joined         object
no_of_dependants      int8
marital_status      object
income               int32
dtype: object

# 05 Merge df_cust dataframe with df_ords_prods_merge

#### Common column between df_cust and df_ords_prods_merge is user_id

In [27]:
#Check user_id data type in df_ords_prods_merge
df_ords_prods_merge['user_id'].dtype

dtype('int64')

In [28]:
#Convert user_id in df_ords_prods_merge to string
df_ords_prods_merge['user_id'] = df_ords_prods_merge['user_id'].astype('str')

In [29]:
df_ords_prods_merge.dtypes

order_id                            int64
user_id                            object
order_number                        int64
order_dow                           int64
order_hour_of_day                   int64
days_since_prior_order            float64
product_id                          int64
add_to_cart_order                   int64
reordered                           int64
product_name                       object
aisle_id                            int64
department_id                       int64
prices                            float64
_merge                           category
price_range_loc                    object
Busiest day                        object
Busiest days                       object
Busiest_hours_of_day               object
max_order                           int64
loyalty_flag                       object
average_price                     float64
spender_flag                       object
median_days_since_prior_order     float64
frequency_flag                    

In [30]:
#Merge df_ords_prods_merge and df_cust
df_ords_prods_cust = df_ords_prods_merge.merge(df_cust, on = 'user_id')

In [31]:
#Checks on merged dataframe
df_ords_prods_cust.shape

(32404859, 33)

In [32]:
df_ords_prods_cust.head()

order_id user_id  order_number  order_dow  order_hour_of_day  \
0   2539329       1             1          2                  8   
1   2398795       1             2          3                  7   
2    473747       1             3          3                 12   
3   2254736       1             4          4                  7   
4    431534       1             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                   999.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...         frequency_flag  first_name  surname  gender  \
0         Soda  ...  Non-frequent customer       Linda   Nguyen  Female   
1         Soda  ...  Non-frequent customer       Linda   Nguyen  Female   
2         Soda  ...  Non-frequent customer       Linda   Nguyen  Female   
3         Soda  ...  Non-frequent customer       Linda   Nguyen  Female   
4         Soda  ...  Non-frequent customer       Linda   Nguyen  Female   

     state age date_joined no_of_dependants  marital_status income  
0  Alabama  31   2/17/2019                3         married  40423  
1  Alabama  31   2/17/2019                3         married  40423  
2  Alabama  31   2/17/2019                3         married  40423  
3  Alabama  31   2/17/2019                3         married  40423  
4  Alabama  31   2/17/2019                3         married  40423  

[5 rows x 33 columns]

# 07 Exporting data

In [33]:
df_ords_prods_cust.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'full_instacart_data.pkl'))